include/dca/phys/dca_step/cluster_solver/shared_tools/accumulation/tp/tp_accumulator_gpu.hpp

```diff
@@ -450,15 +493,90 @@ void TpAccumulator<Parameters, linalg::GPU>::finalize() {
     return;
 
   for (std::size_t channel = 0; channel < G4_.size(); ++channel)
+  {
+    if(distributed_g4_enabled_)
+    {
+      // modify G4 size in G4 cpu, otherwise, copyTo() operation failed due to incomparable size
+      // reset_size() only modifies member Nb_elements in function, does not change tp_dmn.get_size()
+      G4_[channel].reset_size(get_G4()[channel].size());
+    }
     get_G4()[channel].copyTo(G4_[channel]);
-
+  }
   // TODO: release memory if needed by the rest of the DCA loop.
   // get_G4().clear();
 
   finalized_ = true;
   initialized_ = false;
 }
 
+template <class Parameters>
+void TpAccumulator<Parameters, linalg::GPU>::ringG(float& flop) {
```

Comment:
To be reviewed in detail in future.

Reply:
Sure, I can walk you through the algorithm if needed.